src/test_merkle_constructors.rs

```diff
@@ -0,0 +1,338 @@
+#![cfg(test)]
+
+use crate::hash::{Algorithm, Hashable};
+use crate::merkle::{
+    get_merkle_tree_len_generic, Element, FromIndexedParallelIterator, MerkleTree,
+};
+use crate::store::{Store, VecStore};
+use crate::test_common::{Item, Sha256Hasher, XOR128};
+use rayon::iter::{IntoParallelIterator, ParallelIterator};
+use typenum::{Unsigned, U0, U2, U3, U5};
+
+/// Dataset generators. It is assumed that every generator will produce dataset with particular length, equal to leaves parameter
+fn generate_vector_of_elements<E: Element>(leaves: usize) -> Vec<E> {
+    let mut dataset = Vec::<E>::new();
+    for index in 0..leaves {
+        // we are ok with usize -> u8 conversion problems, since we need just predictable dataset
+        let vector: Vec<u8> = (0..E::byte_len()).map(|x| (index + x) as u8).collect();
+        let element = E::from_slice(vector.as_slice());
+        dataset.push(element);
+    }
+    dataset
```

Comment:
I'd also write the outer loop with an functional-style iterator. This way you don't need mutability and it's easy to spot that you operate on all elements. I would write this function as:

```rust
    (0..leaves).map(|index| {
        // we are ok with usize -> u8 conversion problems, since we need just predictable dataset
        let vector: Vec<u8> = (0..E::byte_len()).map(|x| (index + x) as u8).collect();
        E::from_slice(vector.as_slice())
    })
```

Reply:
Added this refactoring. Thanks